tests/i18n/test_compilation.py

```diff
@@ -195,6 +195,66 @@ def test_ignores_based_on_pattern(self):
         self.assertNoneExist(self.CACHE_DIR, ["en", "fr", "it"])
         self.assertNoneExist(self.NESTED_DIR, ["en", "fr", "it"])
 
+    def test_no_dirs_accidentally_skipped(self):
+        """
+        compilemesages uses os.walk to iterate through directories, which yields a
+        3-tuple (dirpath, dirnames, filenames).  When a dirname
+        matching the ignore parameter is found, it is removed from the dirnames list
+        so that os.walk will not recurse into that directory, a feature provided
+        by os.walk: https://docs.python.org/3/library/os.html#os.walk
+
+        However, if our code is iterating through dirnames directly,
+        instead of a copy of it, then removing an 'ignored' directory from the list
+        will cause the dirname directly after the one removed to be skipped.  This will
+        happen any time you remove the current element from a list that you are
+        iterating through, it is not specific to os.walk.
+
+        To recap, even though os.walk continue to recurse into that directory,
+        the loop in which it was skipped will not visit it, which in our case means
+        that a 'locale' directory appearing directly after a skipped directory will be
+        accidentally skipped.
+
+        To actually test this behavior is a bit tricky because os.walk (which uses
+        os.scandir) returns directories in an order that depends on the underlying OS
+        filesystem, and which the python docs refer to as 'arbitrary'.
+        https://docs.python.org/3/library/os.html#os.scandir
+
+        The order IS stable however, in that as long as no other file or directory
+        is created the returned results will be the same across multiple calls.
+
+        Therefore, in order to ensure that our test directory reliably appears before
+        the 'locale' we need to create it first with a carefully constructed name
+        and hope that it will appear first.  As it may not appear first on every OS, the
+        test is skipped intentionally if that is the case.
+        """
+        # Generate the following directory structure
+        # no_accidental_skip/
+        #   _._skipme/
+        #   locale/
+        base_dir = Path("no_accidental_skip")
+        skip_dir = base_dir / "_._skipme"
+        skip_dir.mkdir(parents=True)
+        locale_dir = base_dir / "locale"
+        copytree("canned_locale", locale_dir)
+
+        # Check that the '_._skipme' directory comes just before the 'locale' directory
+        dirlist = [x.name for x in os.scandir(base_dir)]
+        if dirlist.index("locale") - dirlist.index(skip_dir.name) != 1:
+            raise SkipTest(
+                "Could not generate a directory that appears just before `locale`, "
+                "cannot perform test."
+            )
```

Comment:
Yeah this is a bit tricky innit. Mariusz & Natalia will have the final say but this may just be one of the rare times I recommend mocking in order to guarantee a deterministic test 🤔

Reply:
Yeah, I notice that it Skipped when run on the windows CI check, meaning it couldn't create a folder than came before locale.   Not sure if there is some other name trick that would reliably come first on windows.  

I could mock it, but at some point I'm essentially just asserting that this is how lists work in python:

```python
In [1]: x=[1,2,3,4,5,6,7,8,9,10]

In [2]: for i in x:
   ...:      print(i)
   ...:      x.remove(i)
   ...: 
1
3
5
7
9
```

To be honest, I feel like I'm attempting to explain and write tests to prove something that is essentially a fundamental programming mistake (a common and easy one to make).  The effort involved feels like it exceeds the 'risk' of this kind of patch in my opinion.  It is an obvious bug, I have proved that it does in fact cause real world consequences, the solution is trivial and requires no change to the actual logic of the function.  

I don't think it's enormously worth having somewhat fragile tests that spend a good deal of effort proving why you shouldn't remove items from the list you are iterating through.  Likewise, you are having to read and review all of this to approve a safe 6 character obvious fix to an obvious problem.  The sort of thing that a linter rule could probably provide an auto fix for (if the pylint rule actually worked).  That said, I'm happy to make whatever reasonable changes to this PR are required.

Reply:
It took me a while but here is a test that I find readable while ensuring exactly that the whole list of dirnames is traversed (removing the call to `list()` makes the test fail). I'm not proud of using 3 mock calls, perhaps we could create the relevant folder to avoid mocking `isdir`, but the other two IMHO are necessary to keep the test simple-ish:

(EDIT: I re-worked the test a bit after submitting this msg, so the test in the GH email will differ from this one)

```python
    def test_no_dirs_accidentally_skipped(self):
        os_walk_results = [
            # First os.walk call, used to build `basedirs` searching for `locale` dirs
            # and ignoring paths based on `ignore_patterns`
            [("somedir", ["ignore", "locale", "ignore"], [])],
        ]

        locales = ["en", "fr", "it"]
        os_walk_results.extend(
            # Remaining three os.walk calls, used from within the loop iterating over
            # `basedirs`, which in this scenario has 3 entries:
            # exclude/locale, exclude/conf/locale, exclude/somedir/locale
            [(f"dirpath-{d}", [], [f"{locale}.po"])]
            for d in range(3)
            for locale in locales
        )

        import_module = "django.core.management.commands.compilemessages"
        with mock.patch(f"{import_module}.os.walk", side_effect=os_walk_results):
            with mock.patch(f"{import_module}.os.path.isdir", return_value=True):
                with mock.patch(
                    f"{import_module}.Command.compile_messages"
                ) as mock_compile_messages:
                    call_command("compilemessages", ignore=["ignore"], verbosity=4)

        expected = [
            mock.call([(f"dirpath-{d}", f"{locale}.po") for locale in locales])
            for d in range(3)
        ]
        self.assertEqual(mock_compile_messages.mock_calls, expected)
```
@shangxiao @cordery Opinions?

Reply:
Nice 👍  💯 re "the key to the test that we need to add is ensuring that if someone accidentally removes the list() call, a test barfs"

Mocking seems like the only way to achieve that here. We're always asking people to avoid mocking so this is one of those exceptional "Although practicality beats purity." type moments.

Aside from that, this is just my personal opinion but I usually recommend to colleagues that meta-programming be avoided in tests when you can simply statically declare the fixtures as it can improve readability (unless they're too big to read):

```python
        os_walk_results = [
            # First os.walk call, used to build `basedirs` searching for `locale` dirs
            # and ignoring paths based on `ignore_patterns`
            [("somedir", ["ignore", "locale", "ignore"], [])],
            # Remaining three os.walk calls, used from within the loop iterating over
            # `basedirs`, which in this scenario has 3 entries:
            # exclude/locale, exclude/conf/locale, exclude/somedir/locale    
            [('dirpath-1', [], ['en.po'])],
            [('dirpath-1', [], ['fr.po'])],
            [('dirpath-1', [], ['it.po'])],
            [('dirpath-2', [], ['en.po'])],
            [('dirpath-2', [], ['fr.po'])],
            [('dirpath-2', [], ['it.po'])]]
        ]
```
This way people don't need to turn on their brain's Python interpreter to read the test :)

Reply:
> Nice 👍 💯 re "the key to the test that we need to add is ensuring that if someone accidentally removes the list() call, a test barfs"

First of all, sorry for my confusion, I meant that "a test barks" not that "a test barfs" :nauseated_face: 
Though perhaps one could create a somewhat disgusting analogy between a test failure and the latter :rofl: 

> Aside from that, this is just my personal opinion but I usually recommend to colleagues that ~meta-programming~ dynamic fixture construction be avoided in tests when you can simply statically declare the fixtures

I generally agree with this, though there is a delicate balance between readability and making the test as straightforward as possible. In this case, I can't fully decide, defining the lists through expansion will make the test very long :thinking:. I guess we could try to:
* remove the mock for `isdir` and only create the `somedir` folder (so other calls to `isdir` are False and then we'd need to mock less results for `os.walk`)
* have the `os.walk` results defined through expansion

@cordery Would you have time to work on this? If not I'll pick it up soon-ish. Thanks everyone!

Reply:
> First of all, sorry for my confusion, I meant that "a test barks" not that "a test barfs" 🤢
> Though perhaps one could create a somewhat disgusting analogy between a test failure and the latter 🤣
lol, I didn't even blink, it seemed like a totally valid expression to use for a test failure!

Sure thing, thank you for doing the lions share already!  You are both heroes.  I'll rework it this evening.